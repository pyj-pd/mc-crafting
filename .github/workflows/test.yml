name: Test Code

on:
  push:
    branches-ignore:
      - main
  pull-request:
    branches:
      - '**'

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@master
      - name: Install Node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18

      - name: Install pnpm
        uses: pnpm/action-setup@v2
        with:
          version: 7

      - name: Install dependencies and test
        run: |
          pnpm install
          pnpm run test:ci
