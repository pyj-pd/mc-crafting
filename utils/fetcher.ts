import axios from 'axios'
// AXIOS SUCKS

export interface FetcherArgs {
  url: string
  params?: unknown
}

/**
 * Fetcher used on useSWR hook
 */
const fetcher = async ({ url, params = {} }: FetcherArgs) => {
  const res = await axios.get(url, { params }).then((res) => res.data)

  return res
}

export default fetcher
