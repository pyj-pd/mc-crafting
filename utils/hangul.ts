import hangul from 'hangul-js'

interface KeyType {
  [key: string]: string
}

const JAUM: KeyType = {
  ㄱ: 'r',
  ㄴ: 's',
  ㄷ: 'e',
  ㄹ: 'f',
  ㅁ: 'a',
  ㅂ: 'q',
  ㅅ: 't',
  ㅇ: 'd',
  ㅈ: 'w',
  ㅊ: 'c',
  ㅋ: 'z',
  ㅌ: 'x',
  ㅍ: 'v',
  ㅎ: 'g',

  ㄲ: 'R',
  ㄸ: 'E',
  ㅃ: 'Q',
  ㅆ: 'T',
  ㅉ: 'W',

  ㄳ: 'rt',
  ㄵ: 'sw',
  ㄶ: 'sg',
  ㄺ: 'fr',
  ㄻ: 'fa',
  ㄼ: 'fq',
  ㄽ: 'ft',
  ㄾ: 'fx',
  ㄿ: 'fv',
  ㅀ: 'fg',
  ㅄ: 'qt',
}

const MOEUM: KeyType = {
  ㅏ: 'k',
  ㅑ: 'i',
  ㅓ: 'j',
  ㅕ: 'u',
  ㅗ: 'h',
  ㅛ: 'y',
  ㅜ: 'n',
  ㅠ: 'b',
  ㅡ: 'm',
  ㅣ: 'l',

  ㅐ: 'o',
  ㅒ: 'O',
  ㅔ: 'p',
  ㅖ: 'P',
  ㅘ: 'hk',
  ㅙ: 'ho',
  ㅚ: 'hl',
  ㅝ: 'nj',
  ㅞ: 'np',
  ㅟ: 'nl',
  ㅢ: 'ml',
}

/**
 * Convert Hangul to English
 *
 * - `간` -> `rks`
 * - `대` -> `eo`
 * @param text Text to convert
 * @returns Converted text in string
 */
export const hangulToEnglish = (text: string) => {
  const disassembled = hangul.d(text)
  const converted: string[] = disassembled.map((t) => JAUM[t] ?? MOEUM[t] ?? '')

  const result = converted.join('')
  return result
}
