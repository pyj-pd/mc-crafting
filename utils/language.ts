import { readdir, readFile } from 'fs/promises'
import path from 'path'
import { Language, Translation } from 'types/language'

const ALLOWED_PREFIX = ['item', 'block']

const languageListDir = path.resolve(process.cwd(), 'data', 'lang')

/**
 * Get language list
 * @returns Array that contains name and code of each language
 */
export const getLanguageList = async (): Promise<Language[]> => {
  const languageFiles = await readdir(languageListDir)

  const languages: Language[] = []

  for (const languageFile of languageFiles) {
    const languageData = JSON.parse(
      await readFile(path.join(languageListDir, languageFile), 'utf-8')
    )

    const name = languageData['language.name'] as string
    const code = languageFile.split('.')[0]

    languages.push({ name, code })
  }

  return languages
}

/**
 * Get translation list of a language
 * @param code Language code
 * @returns Translations for the language
 */
export const getTranslation = async (code: string): Promise<Translation> => {
  // Read translation file
  const languageFile = JSON.parse(
    await readFile(path.join(languageListDir, `${code}.json`), 'utf-8')
  )

  const data: Translation = {}

  for (const key in languageFile) {
    const split = key.split('.') // item.minecraft.iron_sword

    if (ALLOWED_PREFIX.includes(split[0])) {
      // If the translation is for an item or a block.

      const keyName = split.slice(1).join(':') // minecraft:iron_sword
      data[keyName] = languageFile[key]
    }
  }

  return data
}
