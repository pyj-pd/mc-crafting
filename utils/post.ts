import { readFile } from 'fs/promises'

/**
 * Load post file
 * @param file
 * @returns
 */
export const loadPost = async (file: string) => {
  const fileData = await readFile(file, 'utf-8')

  return fileData
}
