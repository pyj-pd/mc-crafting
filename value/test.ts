// All of these data are for testing only

import { RecipeDataQuery } from 'types/recipe-data'

export const ITEM_ID = 'minecraft:white_wool'
export const TAG_ID = 'minecraft:acacia_logs'

export const LANGUAGE_CODE = 'ko'

export const POST_FILE = 'posts/about.md'

export const CRAFTING_RECIPE_ITEMS: Record<
  'shaped' | 'shapeless',
  RecipeDataQuery
> = {
  shaped: { id: ITEM_ID },
  shapeless: { file: 'light_blue_bed_from_white_bed.json' },
}
