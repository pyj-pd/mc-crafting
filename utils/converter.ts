import path from 'path'
import { readFile, readdir } from 'fs/promises'
import {
  ItemFile,
  ItemNamePair,
  RawRecipeData,
  RawTag,
  TagData,
} from 'types/recipe-data'
import { minecraftPrefix } from 'value/minecraft'
import textures from 'minecraft-textures/dist/textures/1.19'
import { getTranslation } from './language'
import { DEFAULT_LANGUAGE_CODE } from 'value/language'
import { craftingShapedId, craftingShapelessId } from 'value/recipe-data'
import hangul from 'hangul-js'
import Inko from 'inko'

/**
 * Disassemble Korean characters and make it one string
 * @param text Text to disassemble
 * @returns Disassembled text in string
 */
export const disassembleHangul = (text: string): string =>
  hangul.d(text).join('')

const inko = new Inko()

/**
 * Convert Korean characters into English
 * @param text Text to convert
 * @returns Converted text
 */
export const koreanToEnglish = (text: string): string => inko.ko2en(text)

/**
 * Convert text to work well with searching
 * @param text Text to convert
 * @returns Converted text
 */
export const textToSearch = (text: string): string =>
  koreanToEnglish(disassembleHangul(text))

const pureIdRegex = new RegExp(`^${minecraftPrefix}`)

/**
 * Get pure ID of item
 * @param id Item ID
 * @returns Item ID without the prefix
 */
const getPureId = (id: string) => id.replace(pureIdRegex, '')

const recipeListDir = path.resolve(process.cwd(), 'data', 'recipes')
const tagListDir = path.resolve(process.cwd(), 'data', 'tags')

/**
 * Get list of recipe items
 * @returns Item IDs in array
 * @deprecated Use `getItemFiles`
 */
export const getRecipeList = async () => {
  const files = await readdir(recipeListDir)

  return files
}

/**
 * Get ID and file name of items
 * @returns Array that contains items' ID and their file name
 */
export const getItemFiles = async (): Promise<ItemFile[]> => {
  const files = await readdir(recipeListDir)

  const data: ItemFile[] = []

  for (const file of files) {
    const fileData = JSON.parse(
      await readFile(path.join(recipeListDir, file), 'utf-8')
    )

    const type = fileData.type
    const isRecipe =
      (type === craftingShapedId || type === craftingShapelessId) &&
      typeof fileData.result.item === 'string'

    if (!isRecipe) continue

    const id = fileData.result.item

    const item: ItemFile = {
      id,
      file,
    }

    data.push(item)
  }

  return data
}

const itemFiles = await getItemFiles()

/**
 * Item list with recipes
 * @param code Language code
 * @param convert Whether to convert names. Used for searching.
 * @returns Array that contains id, name, and texture of each item
 */
export const getItemList = async (
  code = DEFAULT_LANGUAGE_CODE,
  convert = false
): Promise<ItemNamePair[]> => {
  const translation = await getTranslation(code)

  const itemNames: ItemNamePair[] = []

  for (const { id, file } of itemFiles) {
    const itemData = textures.items.find((item) => item.id === id)

    if (!itemData) continue

    const { texture } = itemData

    const name = translation[id]
    const searchName = convert ? textToSearch(name) : name

    const item: ItemNamePair = {
      id,
      file,
      name,
      searchName,
      texture,
    }

    itemNames.push(item)
  }

  return itemNames
}

const itemList = await getItemList()

/**
 * Get crafting recipe
 * @param file Item file name
 * @param id Item ID
 * @returns Crafting recipe of that item
 */
export const getRecipeData = async ({
  file,
  id,
}: {
  file?: string
  id?: string
}) => {
  const item = itemFiles.find((item) =>
    file ? item.file === file : item.id === id
  )

  if (!item) throw new Error('No recipe.')

  const fileDir = path.join(recipeListDir, item.file)

  const data = JSON.parse(await readFile(fileDir, 'utf-8')) as RawRecipeData

  return data
}

/**
 * Get image of the item
 * @param id Item ID
 */
export const getImage = (id: string) =>
  textures.items.find((item) => item.id === id)?.texture ?? null

/**
 * Get list of item and block in a tag
 * @param id Tag ID
 * @returns Array that contains item or block id
 */
export const getTagData = async (id: string): Promise<TagData> => {
  const tagPureId = getPureId(id)
  const tagFileName = `${tagPureId}.json`
  const tagTypes = await readdir(tagListDir)

  let tagFileDir: string | null = null

  for (const dir of tagTypes) {
    // Search through tag files
    const tagList = await readdir(path.join(tagListDir, dir))
    if (tagList.includes(tagFileName)) {
      tagFileDir = path.join(tagListDir, dir, tagFileName)
      break
    }
  }

  if (tagFileDir === null) throw new Error('No such file.')

  const tagFile = JSON.parse(await readFile(tagFileDir, 'utf-8')) as RawTag

  const data: TagData = tagFile.values

  return data
}

/**
 * Check if crafting recipe for the item exists
 * @param id Item ID
 * @returns boolean
 * @deprecated
 */
export const checkRecipeExists = async (id: string) => {
  const exists = itemList.findIndex((item) => item.id === id) !== -1

  return exists
}
