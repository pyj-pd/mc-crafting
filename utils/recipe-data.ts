import {
  CraftingShaped,
  CraftingShapeless,
  RawRecipeData,
  RecipeData,
} from 'types/recipe-data'
import { getImage, getTagData } from './converter'
import { craftingShapelessId } from 'value/recipe-data'

/**
 * Get each texture for items
 * @param ids Array of item IDs
 * @returns Array of item ID and its texture
 */
const getTextures = (ids: string[]) =>
  ids.map((id) => ({
    id,
    texture: getImage(id),
  }))

type ConvertedData = Pick<RecipeData, 'key' | 'pattern' | 'textures'>

const keyNames = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i']
const convertShapelessData = async (
  data: CraftingShapeless
): Promise<ConvertedData> => {
  const {
    ingredients,
    result: { item },
  } = data

  const items: string[] = [item] // Used to add textures

  // Convert ingredients into keys
  const key: ConvertedData['key'] = {}

  for (const [index, item] of ingredients.entries()) {
    const keyName = keyNames[index] // Key name to be used in pattern
    const keyData = item // Key data with item or tag

    const keyItems = []

    if (keyData.item) {
      items.push(keyData.item)
      keyItems.push(keyData.item)
    } else if (keyData.tag) {
      // If it's tag instead of single item,
      // add all items into texture list

      const tagItems = await getTagData(keyData.tag)

      tagItems.forEach((i) => {
        items.push(i)
        keyItems.push(i)
      })
    }

    key[keyName] = keyItems
  }

  const textures = getTextures(items)

  // Make patterns
  const pattern = Object.keys(key)

  return {
    key,
    pattern,
    textures,
  }
}

const convertShapedData = async (
  data: CraftingShaped
): Promise<ConvertedData> => {
  const {
    key: rawKey,
    result: { item },
  } = data

  const items: string[] = [item] // Used to add textures

  const key: ConvertedData['key'] = {}

  for (const keyName in rawKey) {
    const keyData = rawKey[keyName] // Key data with item or tag
    const keyItems = []

    if (keyData.item) {
      items.push(keyData.item)
      keyItems.push(keyData.item)
    } else if (keyData.tag) {
      // If it's tag instead of single item,
      // add all items into texture list

      const tagItems = await getTagData(keyData.tag)

      tagItems.forEach((i) => {
        items.push(i)
        keyItems.push(i)
      })
    }

    key[keyName] = keyItems
  }

  const textures = getTextures(items)

  const pattern = data.pattern.map((row) => row.padEnd(3, ' ').split('')).flat()

  return { key, pattern, textures }
}

/**
 * Converts raw recipe data into key & pattern data
 * @param data Raw crafting recipe data
 * @returns Converted recipe data
 */
export const convertRecipeData = async (
  data: RawRecipeData
): Promise<RecipeData> => {
  const toAdd = {
    count: data.result.count ?? 1,
    id: data.result.item,
  }

  const id = data.type

  if (id === craftingShapelessId) {
    // Shapeless data
    const converted = await convertShapelessData(data as CraftingShapeless)

    return {
      ...toAdd,
      ...converted,
    }
  } else {
    // Shaped data
    const converted = await convertShapedData(data as CraftingShaped)

    return {
      ...toAdd,
      ...converted,
    }
  }
}
