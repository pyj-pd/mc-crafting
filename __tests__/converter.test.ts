import { ITEM_ID, TAG_ID } from 'value/test'
import {
  getImage,
  getItemFiles,
  getItemList,
  getRecipeData,
  getTagData,
} from '../utils/converter'
import { hangulToEnglish } from '../utils/hangul'
import { describe, it, expect } from 'vitest'

const HANGUL_TO_ENGLISH = {
  input: '갅닮안',
  result: 'rkswekfadks',
}

describe('Data converter', () => {
  it('converts Hangul to English', () => {
    const { input, result } = HANGUL_TO_ENGLISH

    const test = hangulToEnglish(input)

    expect(test).toBe(result)
  })

  it('loads item files', async () => {
    const itemFiles = await getItemFiles()

    expect(itemFiles.length).toMatchInlineSnapshot('754')
  })

  it('loads recipe', async () => {
    const recipeData = await getItemList()

    expect(recipeData.length).toMatchInlineSnapshot('754')
  })

  it('loads recipe of an item', async () => {
    const recipeData = await getRecipeData({ id: ITEM_ID })

    expect(recipeData).toMatchSnapshot()
  })

  it('loads item image', () => {
    const image = getImage(ITEM_ID)

    expect(image).not.toBeNull()
  })

  it('loads tag data', async () => {
    const tagData = await getTagData(TAG_ID)

    expect(tagData).not.toBeFalsy()
  })
})
