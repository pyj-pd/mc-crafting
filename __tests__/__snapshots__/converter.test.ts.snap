// Vitest Snapshot v1

exports[`Data converter > loads recipe of an item 1`] = `
{
  "category": "building",
  "key": {
    "#": {
      "item": "minecraft:string",
    },
  },
  "pattern": [
    "##",
    "##",
  ],
  "result": {
    "item": "minecraft:white_wool",
  },
  "type": "minecraft:crafting_shaped",
}
`;
